### ISSUE TITLE
Notes on using Lonboard in a remote (esp JupyterHub) environment

### ISSUE THREAD

**Issue by kylebarron** [developmentseed/lonboard#397]
I was pairing with @wildintellect on debugging why Lonboard wasn't working in a JupyterHub environment and we came up with a couple notes.

In particular, it's very easy to get lost in environments. Generally a JupyterHub server has _at least two_ Python environments. There's a _server_ environment in which JupyterHub itself is installed and from which it's launched, plus one or more user _client_ environments. As far as I can tell, a Jupyter widget needs to be installed in **both** environments, and presumably those versions must match. 

The widget installed on the _server_ environment is necessary to provide the JS bundle as part of the webpage the user loads, while the widget installed on the _user_ environment is necessary so the Python model works.

- Are there ways to provide helpful errors when no JS code is found? Only a JS console error is produced. Is there some way to catch that and provide a more helpful error to the user?
- Is there a good way to validate that the frontend and backend versions of lonboard match?

Aside from lonboard-specific tasks, I wonder if there are tasks that JupyterHub could do that would simplify the loading of widgets so that they _don't_ need to be separately pre-installed on the server. Additionally, ensuring versions match between the user's environment and the global server environment sounds _messy_.

I've noticed that Colab handles this problem quite well (I think it's with their [googlecolab/colab-cdn-widget-manager](https://github.com/googlecolab/colab-cdn-widget-manager) project). In particular, the first time that a widget is attempted to be rendered, the JS code for that widget is _dynamically_ added to the page. This means that

```
!pip install lonboard
from lonboard import viz
viz([])
```
will _just work_, even without restarting the kernel or reloading the browser tab, because the JS side doesn't need to exist already when the Python side is created.

cc @batpad

**Comment by paddymul** [comment]
FWIW In my own IPYwidget library buckaroo, I addressed some of this.  
Users frequently pip installed the library without restarting the server.  This causes a JS error with no python error.   I added a check on the MTime of the python library, and the start time of the server.  


https://github.com/paddymul/buckaroo/blob/main/buckaroo/widget_utils.py#L37-L55

I'm interested to hear about other approaches around this problem.  It's a really poor experience for an initial user install.  Maybe you could source the JS from unpackage when the user would encounter the problem.

**Comment by ymoisan** [comment]
> discussion on this before. See especially https://github.com/manzt/anywidget/issues/385#issuecomment-1821263446.

says "... a virtual environment (Python venv, Conda environment or what have you) where package jupyterlab has been deployed — **one deploys their widget package in the same environment as where Jupyter Lab lives**" (bold mine).

When I was running for example your NYC building notebook from a Jupyter Lab instance launched from my `jupyterlab` conda env -- to which I had made my `lonboard` conda env visible as a kernel with the usual ipykernel dance -- I was having rendering issues e.g.

```Python
[1] m
Map(layers=[PolygonLayer(extruded=True, get_elevation=<pyarrow.lib.FloatArray object at 0x0000028B03972680>
[
…
```

And similar errors for ScatterPlot layers and `viz([])`

In order to fix this, I just installed jupyterlab **directly** in my lonboard env, launched JLab from that env and bingo.

Unless there's a "JS dance" to expose the JS of conda envs in kernels exported to jupyterlab, I think that means we need to duplicate jupyterlab installs for those conda envs where there is JS.

lonboard v. '0.9.3' in Jupyter Lab 4.2.1

**Comment by kylebarron** [comment]
> I just installed jupyterlab **directly** in my lonboard env, launched JLab from that env and bingo.

FWIW this is what I usually do personally.

For other cases, all you need is to install _anywidget_ in the JupyterLab environment, so that anywidget's JS code exists in JupyterLab when you start the page. Then you can install `lonboard` separately inside a user environment.

**Comment by kannes** [comment]
This seems to be similar to what happens with `uv`'s `--with jupyter`:

I thought this would set me up with a working environment:
```
$ uv venv
$ uv pip install "apache-sedona[db]" lonboard
$ uv run --with jupyter jupyter lab
```

But running this small example in a notebook:
```
import lonboard
import sedona.db

sd = sedona.db.connect()

lonboard.viz(sd.sql("SELECT ST_SetSRID(ST_Point(0, 0), 4326) as geom"))
```

gave the error `Failed to load model class 'AnyModel' from module 'anywidget'`.

The solution was to explicitly install jupyterlab in the environment:

```
$ uv venv
$ uv pip install "apache-sedona[db]" lonboard jupyterlab
$ uv run jupyter lab
```

This way, `lonboard.viz()` works fine.

**Comment by kylebarron** [comment]
Yes, you need to install anywidget into the environment where JupyterLab is launched from
